File: src/main/java/nucleon/event/Event.java
package nucleon.event;

public abstract class Event {

    private static final int RECURSION_DEPTH_LIMIT = 64;

    private final ThreadLocal<Integer> recursionDepth = ThreadLocal.withInitial(() -> 0);

    private boolean cancelled = false;

    public final void call() {
        int recursionDepthValue = recursionDepth.get();

        if (recursionDepthValue >= RECURSION_DEPTH_LIMIT) {
            throw new EventException("Recursion depth exceeded the limit value (" + RECURSION_DEPTH_LIMIT + ")");
        }

        try {
            recursionDepth.set(recursionDepthValue + 1);

            FilterChain<?> list = FilterChainManager.GLOBAL.acquireChainFor(getClass());

            do {
                list.execute(this);
                list = list.getParent();
            } while (list != null);
        } finally {
            recursionDepth.set(recursionDepthValue);
        }
    }

    public final void cancel() {
        ensureThatCancellable();
        cancelled = true;
    }

    private void ensureThatCancellable() {
        if (!(this instanceof Cancellable)) {
            throw new IllegalStateException("Attempt to cancel non-cancellable event");
        }
    }

    public final boolean isCancelled() { return cancelled; }

    public final String name() { return getClass().getSimpleName(); }
}

File: src/main/java/nucleon/event/Event.kt
package nucleon.event

abstract class Event(
    private val target: EventTarget? = null
) {

    private companion object {
        val RECURSION_DEPTH: ThreadLocal<Int> = ThreadLocal.withInitial { 0 }
        const val RECURSION_DEPTH_LIMIT = 64
    }

    var cancelled = false
        private set

    val name: String = javaClass.simpleName

    fun call() {
        val recursionDepthVal = RECURSION_DEPTH.get()

        if (recursionDepthVal >= RECURSION_DEPTH_LIMIT) {
            throw EventException("Recursion depth exceeded the limit value")
        }

        try {
            RECURSION_DEPTH.set(recursionDepthVal + 1)
            execute()
        } finally {
            RECURSION_DEPTH.set(recursionDepthVal)
        }
    }

    private fun execute() {
        var chain: FilterChain<*> = FilterChainManager.acquireChainFor(javaClass)

        while (true) {
            chain.execute(this)
            chain = chain.parent ?: break
        }

        if (!cancelled) {
            target?.execute()
        }
    }

    fun cancel() {
        check(this is Cancellable) { "Attempt to cancel a non-cancellable event" }
        cancelled = true
    }
}